pkg/resolver/resolver.go

```diff
@@ -0,0 +1,341 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolver
+
+import (
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/pkg/envoy/envoy/api/v2/core"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/resolver/sort"
+)
+
+var log = logging.DefaultLogger
+
+// ResolveIdentityPolicy generates the cilium.NetworkPolicy which corresponds to
+// the policy that applies to the given identityToResolve from the set of rules.
+func ResolveIdentityPolicy(rules api.Rules, identityCache identity.IdentityCache, identityToResolve identity.NumericIdentity) *cilium.NetworkPolicy {
+	log.Infof("resolving policy for identity %d", identityToResolve)
+	identityLabels, ok := identityCache[identityToResolve]
+	if !ok {
+		log.Warningf("no labels corresponding to identity %d exist in the identity cache: %s", identityToResolve, identityCache)
+		return nil
+	}
+
+	// List which tracks rules which select the identity for which policy is
+	// being resolved. This is done because the rules must be passed over twice,
+	// and there's no point in traversing over the same set of rules multiple
+	// times if we can detect on the first pass if the rule selects the identity.
+	selectingRules := api.Rules{}
+
+	// Sets of identities which track identities that are denied upon ingress
+	// and egress, and identities which are not denied (i.e., not members of the
+	// sets of denied identities). The union of both sets for ingress and egress
+	// should always equal the set of identities provided in the identity cache.
+	notDeniedIngressIdentities := identity.IdentityCache{}
+	deniedIngressIdentities := identity.IdentityCache{}
+	notDeniedEgressIdentities := identity.IdentityCache{}
+	deniedEgressIdentities := identity.IdentityCache{}
+
+	identityPolicy := cilium.NetworkPolicy{
+		Name:   "",
```

Comment:
Note that this is empty because "Name" usually corresponds to an IP address of a `NetworkPolicy` (see `pkg/envoy`). But, at this layer, we are resolving per identity, not per endpoint. I should add this as a comment.

Reply:
Noted that Name is not populated and it's up to caller of function to populate it accordingly.